### PR TITLE
Add CRC32 combine functionality

### PR DESCRIPTION
**Pull request by tab1293**
Including @justinruggles CRC32 combine functionality from firehose

### CODE REVIEW

File: crc32combine/crc32combine.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    return sum
+}
+
+func gf2MatrixSquare(square, mat []uint) {
+    for n := 0; n < 32; n++ {
+        square[n] = gf2MatrixTimes(mat, mat[n])
+    }
+}
+
+func CRC32Combine(poly uint32, crc1, crc2 uint32, len2 int64) uint32 {
```
Comment (dwbuiten) [inline crc32combine/crc32combine.go]: Missing docs.

File: crc32combine/crc32combine.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    /* degenerate case (also disallow negative lengths) */
```
Comment (dwbuiten) [inline crc32combine/crc32combine.go]: Should use go-style comments.

File: crc32combine/crc32combine.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+   misrepresented as being the original software.
+
+3. This notice may not be removed or altered from any source distribution.
+
+Ported from C to Go in 2016 by Justin Ruggles, with minimal alteration.
+Used uint for unsigned long. Used uint32 for input arguments in order to match
+the Go hash/crc32 package.
+*/
+
+func gf2MatrixTimes(mat []uint, vec uint) uint {
```
Comment (dwbuiten) [inline crc32combine/crc32combine.go]: `go fmt` Should be run on the file.

File: crc32combine/crc32combine.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+}
+
+func CRC32Combine(poly uint32, crc1, crc2 uint32, len2 int64) uint32 {
+    /* degenerate case (also disallow negative lengths) */
+    if len2 <= 0 {
+        return crc1
+    }
+
+    even := make([]uint, 32) /* even-power-of-two zeros operator */
```
Comment (dwbuiten) [inline crc32combine/crc32combine.go]: Why `make` and not an array? Also, any reason for the conversion to `uint`?

File: crc32combine/crc32combine.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
+package crc32combine
```
Comment (dwbuiten) [inline crc32combine/crc32combine.go]: Missing package doc.

**Comment by dwbuiten** [comment]
General query: Does this single static function really need its own package?

**Comment by tab1293** [comment]
@dwbuiten the reason I wanted this to be a separate package was so I could include it in https://github.com/tus/tusd/pull/106. Need this function check the validity of a compose operation in GCS.

**Comment by dwbuiten** [comment]
@tab1293 OK.

File: crc32combine/crc32combine.go (hunk 4 above)
Comment (justinruggles) [inline crc32combine/crc32combine.go]: `uint` is probably the closest thing to `unsigned long` from the zlib source this was ported from
https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/crc32.c#L341-L425

**Comment by dwbuiten** [comment]
LGTM if:

1. The two commits are squashed properly before merge.
2. The function documentation for `CRC32Combine` is updated to proper godoc style (`// CRC32Combine returns ...`).

**Comment by tab1293** [comment]
@dwbuiten Should be good to go 👍

**Comment by dwbuiten** [comment]
Merged as 8cd4c737f091d9317f72b25df78ce6cf869f7d30.

**Pull request closed without merging**